pkg/codegen/pcl/README.md

```diff
@@ -21,6 +21,139 @@ Pulumi supports a number of operations pertaining to PCL programs:
+## PCL programs
+
+PCL programs are comprised of PCL source files which typically have the suffix
+`.pp`. This repository contains a large number of examples -- in particular, a
+large amount of PCL is used to drive [language conformance
+tests](language-conformance-tests). PCL programs support the following concepts:
+
+(pcl-resource)=
+### Resources
+
+*Resources* are defined as [blocks](pcl-lexing-parsing) of type `resource`. They
+correspond to stack resource registrations (see [resource
+registration](resource-registration) and
+[](pulumirpc.ResourceMonitor.RegisterResource) for more). A resource block must
+have exactly two labels -- the first is the resource's name, to be used both as
+a reference to it elsewhere in the program and as its name in the resulting
+Pulumi stack (see the sections in this documentation on [URNs](urns) and
+[resource registration](resource-registration)). The resource's body contains a
+set of attributes, which correspond to the resource's input properties. An
+optional `options` block may be used to specify resource options (e.g.
+`provider`, `protect`, `ignoreChanges`, etc.).
+
+```hcl
+resource "r" "pkg:index:Resource" {
+    name = "r"
+
+    options {
+        protect = true
+    }
+}
+```
+
+(pcl-component)=
+### Components
+
+A *component* in the context of a PCL program is equivalent to the notion of a
+"local" [component resource](component-resources) in a Pulumi program.
+Components are defined as blocks of type `component`. A component block must
+have two labels. The first is the component's name, to be used both as a
+reference to it elsewhere in the program and as its name in the resulting Pulumi
+stack. The second is a path to a directory containing a PCL program (set of
+`.pp` files) that defines the component's implementation. The component's body
+contains a set of attributes, which correspond to the component's input
+properties. An optional `options` block may be used to specify component options
+(e.g. `provider`, `protect`, etc.).
+
+```hcl
+component "c" "./c" {
+    name = "c"
+    value = 42
+}
+```
+
+Program generation for components results in e.g. the generation of local
+component types, such as those extending the `ComponentResource` class in SDKs
+exposing that type.
+
+(pcl-output)=
+### Outputs
+
+*Outputs* are defined as [blocks](pcl-lexing-parsing) of type `output`. They
+correspond exactly to exported stack outputs, such as `pulumi.export` in Python
+or `ctx.Export` in Go, which correspond to
+[](pulumirpc.ResourceMonitor.RegisterResourceOutputs) on the stack resource
+under the hood. Output blocks must have exactly one label, which is the output's
+name. The output's body must contain a single attribute, `value`, which is the
+output's value.
+
+```hcl
+resource "r" "pkg:index:Resource" {
+    name = "r"
+}
+
+output "o" {
+    value = r.name
+}
+```
+
+(pcl-config)=
+### Configuration
+
+*Configuration variables* define the names and schema of values that are
```

Comment:
I was a little confused by "schema" here, should we say "type" instead? I might be missing something though.

```suggestion
*Configuration variables* define the names and type of values that are
```

Reply:
I think I picked this wording because you can also add descriptions etc., so it's kinda letting you shoe-horn in a full "schema" (name, description, type) into the block, if that makes sense. Open to changing it though!

Reply:
That make sense, happy to keep it as "schema"!